### DISCUSSION TITLE
Fields interpolation on lagrangian particle trajectories

### DISCUSSION THREAD

**Discussion by weigwu** [CliMA/Oceananigans.jl#3297] [Computational science!] [answered]
Hi, I am using the LagrangianParticles() function in Oceananigans. I am wondering if there is a way to interpolate the fields (e.g. u, v, w, b) onto particle trajectories and save the fields for each particle.

Many thanks,
Roger

**Comment by jagoosw** [comment, marked as answer]
Hi Roger,

You could do this by making a custom particle, using tracked fields to set the particle properties, and saving the particles the normal way. For example:

```julia
using Oceananigans
using Oceananigans.Fields: VelocityFields
using StructArrays

struct TrackingParticle{FT}
    x :: FT
    y :: FT
    z :: FT
    u :: FT
    v :: FT
    w :: FT
    b :: FT
end

grid = RectilinearGrid(; size = (128, 128, 128), extent = (100, 100, 100))

n = 100

x₀, y₀, z₀, u₀, v₀, w₀, b₀ = [zeros(n) for i in 1:7]

particles = StructArray{TrackingParticle}((x₀, y₀, z₀, u₀, v₀, w₀, b₀))

u, v, w = VelocityFields(grid)

b = CenterField(grid)

lagrangian_particles = LagrangianParticles(particles, tracked_fields = (; u, v, w, b))

model = NonhydrostaticModel(; grid, velocities = (; u, v, w), particles = lagrangian_particles, buoyancy = BuoyancyTracer(), tracers = (; b))

simulation = Simulation(model, Δt = 1, stop_iteration = 100)

simulation.output_writers[:particles] = JLD2OutputWriter(model, (particles=model.particles,), filename="particles", schedule=IterationInterval(1))
```

**Reply by glwagner** [reply]
Worth noting that this is written in the docstring:

```julia
"""
    LagrangianParticles(particles::StructArray; restitution=1.0, tracked_fields::NamedTuple=NamedTuple(), dynamics=no_dynamics)

Construct some `LagrangianParticles` that can be passed to a model. The `particles` should be a `StructArray`
and can contain custom fields. The coefficient of restitution for particle-wall collisions is specified by `restitution`.

A number of `tracked_fields` may be passed in as a `NamedTuple` of fields. Each particle will track the value of each
field. Each tracked field must have a corresponding particle property. So if `T` is a tracked field, then `T` must also
be a custom particle property.

`dynamics` is a function of `(lagrangian_particles, model, Δt)` that is called prior to advecting particles.
`parameters` can be accessed inside the `dynamics` function.
"""
```

However, it's a little terse! It might be nice if the docstring had an example (eg the one written above by @jagoosw).

Also, is it always required to create the fields in advance? Is there any way to implement a design that's a little easier to use?

**Reply by jagoosw** [reply]
As far as I'm aware there isn't a way to do it without creating the fields first (I recall having problems with this when I was getting started too).

It would probably be easiest to change `tracked_fields` to a tuple of symbols, and then change:
https://github.com/CliMA/Oceananigans.jl/blob/e16cdc6cfb67703df9e29368017868331f68b1c0/src/Models/LagrangianParticleTracking/update_lagrangian_particle_properties.jl#L16-L35
to something along the lines of:
```
...
     fields = fields(model)

     for name in particles.tracked_fields
         field = fields[name]
...
```

This does make it slightly more complicated to track something like vorticity, I suppose you could make it an auxiliary field of the model though.

**Reply by glwagner** [reply]
Are `tracked_fields` used in a kernel? Maybe it can be a `Dict` (which can be appended to after a model is created) rather than `NamedTuple`.

Would definitely also be useful to have a way to auto add prognostic and auxiliary model fields!

**Reply by glwagner** [reply]
Also worth noting that, as far as I can tell, `interpolate` works with `AbstractOperations` and so it should be possible to use them as tracked fields (ie the tracked fields do not _have_ to be precomputed).

**Reply by weigwu** [reply]
Hi Jago and Gregory, thanks a lot for providing an example. Do you have any idea in tracking a field involving the derivative of velocities (e.g., vorticity)? Can I still make a custom particle as the way Jago showed?

**Reply by jagoosw** [reply]
> Are `tracked_fields` used in a kernel? Maybe it can be a `Dict` (which can be appended to after a model is created) rather than `NamedTuple`.
> 
> Would definitely also be useful to have a way to auto add prognostic and auxiliary model fields!

I don't think they're used in kernels. I don't know why we would need particles to have their own fields that can't just be part of the model. Wouldn't adding in that way require a materialisation step in the model setup also?

**Reply by jagoosw** [reply]
> Hi Jago and Gregory, thanks a lot for providing an example. Do you have any idea in tracking a field involving the derivative of velocities (e.g., vorticity)? Can I still make a custom particle as the way Jago showed?

You can make an abstract operation and track that, e.g. `ζ = ∂x(v) - ∂y(u)`, then add `ζ` as a property in the `struct` and add to `tracked_fields`.

**Reply by glwagner** [reply]
Note the tracked fields can be any abstract array, so fields, abstract operations, and any other array-like object that supports the 3D index `i, j, k` will work.

**Comment by jagoosw** [comment]
> Are `tracked_fields` used in a kernel? Maybe it can be a `Dict` (which can be appended to after a model is created) rather than `NamedTuple`.
> 
> Would definitely also be useful to have a way to auto add prognostic and auxiliary model fields!

I don't think they're used in kernels. I don't know why we would need particles to have their own fields that can't just be part of the model. Wouldn't adding in that way require a materialisation step in the model setup also?

**Reply by glwagner** [reply]
I remember we had this discussion before also about `auxiliary_fields` --- those would also benefit from being `Dict`. One argument against is that model cannot be concretely typed. I just don't know if that kind of change would have unforseen negative future consequences --- for example, for automatic differentiation of `time_step!(model, dt)`.

**Reply by glwagner** [reply]
I wonder --- should we fix how particles are implemented so that they can be built outside the model? What is the advantage of having them be a property of model? It seems like their dynamics could be handled via callbacks. This would add flexibility and might simplify things.

**Comment by jagoosw** [comment]
> Hi Jago and Gregory, thanks a lot for providing an example. Do you have any idea in tracking a field involving the derivative of velocities (e.g., vorticity)? Can I still make a custom particle as the way Jago showed?

You can make an abstract operation and track that, e.g. `ζ = ∂x(v) - ∂y(u)`, then add `ζ` as a property in the `struct` and add to `tracked_fields`.
